ecosystem/sep-0045.md

```diff
@@ -0,0 +1,431 @@
+Track: Standard
+Status: Draft
+Created: 2024-10-08
+Updated: 2024-10-08
+Version: 0.1.0
+Discussion: TBA
+```
+
+## Simple Summary
+
+This SEP defines the standard way for clients such as wallets or exchanges to create authenticated web sessions on
+behalf of a user who holds a contract account. A wallet may want to authenticate with any web service which requires a
+contract account ownership verification, for example, to upload KYC information to an anchor in an authenticated way as
+described in [SEP-12](sep-0012.md).
+
+This SEP is based on [SEP-10](sep-0010.md), but only supports contract accounts and aims to provide a JWT that is fully
+compatible with existing protocols that use SEP-10.
+
+## Definitions
+
+- **Address**: An identifier representing a possible account on the Stellar network. Addresses begin with specific
+  prefixes like G for public keys (accounts controlled by secret keys), M for muxed accounts and C for contract
+  accounts.
+- **Account**: A stateful entity on the Stellar network that can hold balances. This can refer to a Stellar account
+  (`G...` addresses or `M...` addresses) or Soroban contract that abstracts an account (`C...` addresses).
+
+## Abstract
+
+This protocol is a variation of mutual challenge-response, which uses Soroban authorization entries to encode challenges
+and responses.
+
+It involves the following components:
+
+- A **Home Domain**: a domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `WEB_AUTH_ENDPOINT_SEP0045` (URL)
+  and `SIGNING_KEY` (`G...`).
+  - The `SIGNING_KEY` from the **Home Domain** is the **Home Domain Address**.
+- A **Server**: a server providing the `WEB_AUTH_ENDPOINT_SEP0045` that implements the GET and POST operations discussed
+  in this document. The server's domain may be the **Home Domain**, a sub-domain of the **Home Domain**, or a different
+  domain.
+- A **Client Account**: A Soroban contract that has been deployed to the network which is being authenticated.
+- A **Client**: the software used by the holder of the **Client Account** being authenticated by the **Server**.
+- A **Client Domain** (optional): a domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `SIGNING_KEY` used
+  for [Verifying the Client Domain](#verifying-the-client-domain)
+  - The `SIGNING_KEY` from this domain is the **Client Domain Address**
+- A **Web Auth Contract**: a contract that implements the `web_auth_verify` function as described in this document. The
+  contract must be deployed at the `WEB_AUTH_CONTRACT_ID` address specified in the **Server**'s `stellar.toml`.
+
+The discovery flow is as follows:
+
+1. The **Client** retrieves the `stellar.toml` from the **Home Domain** in accordance with
+   [SEP-1 stellar.toml](sep-0001.md).
+1. The **Client** looks up the `WEB_AUTH_ENDPOINT_SEP0045`, `WEB_AUTH_CONTRACT_ID` and `SIGNING_KEY` (i.e. **Home Domain
+   Address**) from the `stellar.toml`.
+
+The authentication flow is as follows:
```

Comment:
The auth flow in SEP-10 contains a really important step:

> The Client verifies that the transaction has an invalid sequence number 0. This is extremely important to ensure the transaction isn't malicious.

A similar step doesn't appear in the auth flow for SEP-45. Does the same problem get tackled another way?

Reply:
I couldn't think of a way to construct the challenge such that the signed transaction succeeds in simulation but errors when submitted to the network. Instead, I recommend that the client check that the challenge is not malicious by simulating it before returning it to the server.

Reply:
An alternative approach would require `web_auth_verify` to return a specific error code when all signatures are valid. The server would check for this error code when it verifies the client's response.

Reply:
An alternative approach would be to implement `web_auth_verify` to fail with a specific error when all signatures are valid. The server would have to manually construct the authorization entries which can be challenging due to the nonces.